tests/data_context/cloud_data_context/test_datasource_crud.py

```diff
@@ -358,3 +369,281 @@ def test_cloud_data_context_add_datasource(
         assert stored_datasource.name == datasource_name
         assert stored_datasource.config["name"] == datasource_name
         assert stored_data_connector.name == data_connector_name
+
+
+@pytest.mark.unit
+def test_non_cloud_backed_data_context_save_datasource_empty_store(
+    empty_data_context: DataContext,
+    datasource_config_with_names: DatasourceConfig,
+):
+    context: DataContext = empty_data_context
+    # Make sure the fixture has the right configuration
+    assert len(context.list_datasources()) == 0
+
+    datasource: Datasource = context._build_datasource_from_config(
+        datasource_config_with_names
+    )
+    datasource_name: str = datasource.name
+
+    with patch(
+        "great_expectations.data_context.store.datasource_store.DatasourceStore.set",
+        autospec=True,
+        return_value=datasource_config_with_names,
+    ):
+        updated_datasource: Union[
+            LegacyDatasource, BaseDatasource
+        ] = context.save_datasource(datasource)
+
+        # Make sure the datasource config got into the context config
+        assert len(context.config.datasources) == 1
+        assert (
+            context.config.datasources[datasource_name] == datasource_config_with_names
+        )
+
+        cached_datasource = context._cached_datasources[datasource_name]
+
+        # Make sure the datasource got into the cache
+        assert len(context._cached_datasources) == 1
+
+        # Make sure the stored and returned datasource is the same one as the cached datasource
+        assert id(updated_datasource) == id(cached_datasource)
+        assert updated_datasource == cached_datasource
+
+        # Make sure the stored and returned datasource is a different instance
+        assert not id(cached_datasource) == id(datasource)
+        assert not id(updated_datasource) == id(datasource)
+        assert not cached_datasource == datasource
+        assert not updated_datasource == datasource
+
+        # Make sure the stored and returned datasource are otherwise equal
+        serializer: AbstractConfigSerializer = DictConfigSerializer(
+            schema=datasourceConfigSchema
+        )
+        updated_datasource_dict = serializer.serialize(
+            datasourceConfigSchema.load(updated_datasource.config)
+        )
+        orig_datasource_dict = serializer.serialize(
+            datasourceConfigSchema.load(datasource.config)
+        )
+
+        for attribute in ("name", "execution_engine", "data_connectors"):
+            assert updated_datasource_dict[attribute] == orig_datasource_dict[attribute]
+
+
+@pytest.mark.unit
+def test_non_cloud_backed_data_context_save_datasource_overwrite_existing(
```

Comment:
There is a lot of duplicated code between this test and the one above, please factor it out and reference from both tests.

I am not clear from the description what the difference is between this test and the one above

Reply:
This tests the overwrite or "update" functionality by first saving a datasource to the store, then saving another datasource with the same name and a slightly different config. It makes sure there is only 1 datasource in the store and that the config was updated (rather than creating a second datasource).

Reply:
And good call on the duplicated code, coming up in a new commit.

Reply:
So far I've moved some duplicated code into _save_datasource_assertions. Looking for other opportunities.